policykit/policyengine/models.py

```diff
@@ -1026,6 +1027,143 @@ def __init__(self, *args, **kwargs):
+    @property
+    def permissions(self):
+        if self.community_name:
+            # If it is a user custom action, it has a new permission name
+            permissions = ((f"can_execute_{self.community_name}", "Can execute {self.community_name}"))
+        else:
+            # Otherwise, this permission for this new CustomAction is the same as the GovernableAction it builts upon 
+            action_content_type = ContentType.objects.filter(model=self.action_type)     
+            all_permissions = Permission.objects.filter(content_type__in=action_content_type)
+            # Search for all permissions related to this GovernableAction
+            permissions = [(perm.codename, perm.name) for perm in all_permissions if perm.codename.startswith("can_execute") ]
+            
+            # While it is obvious that the permission codename is f"can_execute_{self.action_type}", 
+            # We actually do not know exactly the corresponding permisson name
+            # That is why we take such trouble to extract it
+        return permissions
+
+
+    def save(self, *args, **kwargs):
+        """
+        Add the permission if it is a user custom action
+        """
+
+        if not self.pk and self.community_name:
+            action_content_type = ContentType.objects.get_for_model(CustomAction)
+            all_permissions = self.permissions
+            for perm in all_permissions:
+                # TODO not sure whether we should use the content type of CustomAction here
+                Permission.objects.create(codename=perm[0], name=perm[1], content_type=action_content_type)
+                # TODO not sure we would like to assign this permission to all users by default or not
+                # perhaps we should at first assign it to users who have the execute permission of the referenced GovernableAction
+        super(CustomAction, self).save(*args, **kwargs)
+
+class Procedure(models.Model):
+
+    name = models.TextField(blank=True, default='')
+    """the name of the procedure if exists"""
+
+    initial_code = models.TextField(blank=True, default='')
+
+    notify_code = models.TextField(blank=True, default='')
+
+    check_code = models.TextField(blank=True, default='')
+
+    variables_dict = models.TextField(blank=True, default='')
+        
+    def parse_variables_dict(self):
+        return json.loads(self.variables_dict)
+    
+    def generate_codes(self):
+        """
+            Generate codes for intialize, check, notify, success, and fail based on procedure variables defined in each subclass
+        """
+        variables = self.parse_variables_dict()
+        for variable in variables:
+            PolicyVariable.objects.create(**variable)
+
+        return {
+            "initial": self.initial_code,
+            "notify": self.notify_code,
+            "check": self.check_code,
+        }
+
+class Execution(models.Model):
+    
+    action = models.ForeignKey(GovernableAction, on_delete=models.SET_NULL)
```

Comment:
Do we actually want this to be a foreignkey to an action? Do we want to create an action object whenever we run an execution? What would the purpose of that be (in terms of storing data we want to use later)? I think we want actions at the beginning because they form proposals, and we need the stored data both to carry out the procedure and to be able to execute the action (if need be) when passed. But I don't see the need to create an action object at execution time.

Reply:
At the end of the day, the user is selecting from a menu of action types, and perhaps specifying parameters for that action type. It's similar to CustomAction in fact? We don't need to create the action object, we should just run the code to execute.

Reply:
Then we can keep a reference to the `ActionType` here and put the execution codes of each specific GovernableAction in an execution code block? When I was creating a reference to the GovernableAction, my intention was to have an organized way to store the execution parameters of each Governable action so that we can directly know who is the user, and what message is posted (compared to embedding all these parameters in the rather execution codes of different shapes). But I do not have a clear idea about what is the future use of these organized data and whether it is worth the cost.

Reply:
Hm yes, one way to do this is to create action objects in memory but never actually commit them to the database.

Reply:
Shouldn't we distinguish Pass vs Fail as well in here?

Reply:
I am thinking users can create one Execution instance for Pass and Fail respectively. Putting the PASS and FAIL code blocks together seems to assume that these two are quite related, and also introduces additional logic in the Execution implementation.

Reply:
ok, then do we need a field specifying which one it is?